bigtable/admin/admin_client.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/admin/admin_client.h"
+
+#include <absl/memory/memory.h>
+
+namespace {
+/// An implementation of the bigtable::AdminClient interface.
+class SimpleAdminClient : public bigtable::AdminClient {
+ public:
+  SimpleAdminClient(std::string project, bigtable::ClientOptions options)
+      : project_(std::move(project)), options_(std::move(options)) {}
+
+  std::string const& project() const override { return project_; }
+  void on_completion(grpc::Status const& status) override;
+  ::google::bigtable::admin::v2::BigtableTableAdmin::StubInterface&
+  table_admin() override;
+
+ private:
+  void refresh_credentials_and_channel();
+
+ private:
+  std::string project_;
+  bigtable::ClientOptions options_;
+  mutable std::mutex mu_;
+  std::shared_ptr<grpc::Channel> channel_;
+  std::unique_ptr<
+      google::bigtable::admin::v2::BigtableTableAdmin::StubInterface>
+      table_admin_stub_;
+};
+}  // anonymous namespace
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+std::shared_ptr<AdminClient> CreateAdminClient(
+    std::string project, bigtable::ClientOptions options) {
+  return std::make_shared<SimpleAdminClient>(std::move(project),
+                                             std::move(options));
+}
+
+}  // namespace BIGTABLE_CLIENT_NS
+}  // namespace bigtable
+
+namespace {
+void SimpleAdminClient::on_completion(grpc::Status const& status) {
+  if (status.ok()) {
+    return;
+  }
+  std::unique_lock<std::mutex> lk(mu_);
+  channel_.reset();
+  table_admin_stub_.reset();
+}
+
+::google::bigtable::admin::v2::BigtableTableAdmin::StubInterface&
+SimpleAdminClient::table_admin() {
+  refresh_credentials_and_channel();
+  return *table_admin_stub_;
+}
+
+void SimpleAdminClient::refresh_credentials_and_channel() {
+  std::unique_lock<std::mutex> lk(mu_);
+  if (table_admin_stub_) {
+    return;
+  }
+  // Release the lock before executing potentially slow operations.
```

Comment:
Rather than doing that, would it make sense to create separate blocks, and use `std::unique_lock` and let its destructor automatically unlock, so you don't also have to relock? E.g., something like:

```c++
{
  std::unique_lock lk(mu_);
  // some code under mutex here
}

// some slow code here

{
  std::unique_lock lk(mu_);
  // some more code under mutex here
}
```

It might read better (potentially), you skip the manual unlock/lock steps. Is constructing a new `std::unique_lock` expensive?

Thoughts?

Reply:
Done.